src/sensr_client.cpp

```diff
@@ -24,32 +24,43 @@ namespace sensr
     }
   }
 
-  bool Client::Reconnect() {
-    bool ret1 = false;
-    bool ret2 = false;
-    // Reconnect if no listener is listening OutputMessage.
-    if (std::any_of(listeners_.begin(), listeners_.end(), [](const std::shared_ptr<MessageListener>& listener) {
-      return listener->IsOutputMessageListening();
-    })) {
-      output_endpoint_->Close(websocketpp::close::status::normal);
-      ret1 = output_endpoint_->Connect(GetProtocol() + "://" + address_ + ":5050", 
-        std::bind(&Client::OnResultMessage, this, std::placeholders::_1),
-        std::bind(&Client::OnResultError, this, std::placeholders::_1));
-    } else {
-      ret1 = true;
-    }
-    // Close if no listener is listening PointResult.
-    if (std::any_of(listeners_.begin(), listeners_.end(), [](const std::shared_ptr<MessageListener>& listener) {
-      return listener->IsPointResultListening();
-    })) {
-      point_endpoint_->Close(websocketpp::close::status::normal);
-      ret2 = point_endpoint_->Connect(GetProtocol() + "://" + address_ + ":5051", 
-        std::bind(&Client::OnPointMessage, this, std::placeholders::_1),
-        std::bind(&Client::OnPointError, this, std::placeholders::_1)); 
-    } else {
-      ret2 = true;
+  void Client::Reconnect() {
+    if (!is_reconnecting_) {
+      std::thread t([this]() {
+        is_reconnecting_ = true;
+        auto temp = listeners_;
+        size_t reconnect_count_ = Client::kMaxReconnectTrialCount;
+        while (reconnect_count_-- > 0) {
+          for (const auto& listener : listeners_) {
+            UnsubscribeMessageListener(listener);
+          }
+          for (const auto& listener : temp) {
+            SubscribeMessageListener(listener);
+          }
+          bool ret1 = true;
+          bool ret2 = true;
+          if (IsResultListening()) {
+            if (!output_endpoint_->IsConnected()) {
+              ret1 = false;
+            }
+          }
+          if (IsPointListening()) {
+            if (!point_endpoint_->IsConnected()) {
+              ret2 = false;
+            }
+          }
+
+          if (ret1 && ret2) {
+            INFO_LOG("Reconnected!");
+            break;
+          }
+          std::this_thread::sleep_for(std::chrono::seconds(1));
+          INFO_LOG("Reconnecting...");
+        }
+        is_reconnecting_ = false;
+      });
+      t.detach();
```

Comment:
Is it safe to detach this here? What if we get in an endless loop?

Reply:
Now, re-connection thread is managed by sdk source.